File: app/src/main/java/io/github/lmller/codetalks2017/SearchViewModel.java
package io.github.lmller.codetalks2017;

import android.databinding.BaseObservable;
import android.databinding.Bindable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;
import io.reactivex.Scheduler;
import io.reactivex.disposables.Disposable;


public class SearchViewModel extends BaseObservable {
	private final Thesaurus synonyms;

	private boolean isLoading;
	private List<Term> searchResults = Collections.emptyList();
	private String errorMessage;
	private String logTag;

	public SearchViewModel(Thesaurus synonymSource) {
		this.synonyms = synonymSource;
		this.logTag = synonymSource.name();
	}

	public Disposable observeTextChanges(Observable<String> queryTextChange, Scheduler scheduler) {
		return queryTextChange
				.filter(text -> !text.isEmpty())
				.debounce(250, TimeUnit.MILLISECONDS, scheduler)
				.flatMap(text -> synonyms.findSynonym(text)
						.doOnError(e -> setErrorMessage(e))
						.doOnSubscribe(s -> resetState())
						.doOnTerminate(() -> isLoading = false)
						.onErrorReturnItem(new Synset())
				)
				.subscribe(
						result -> handleSearchResult(result)
				);
	}

	private void setErrorMessage(Throwable e) {
		errorMessage = logTag + " responded with: " + e.getMessage();
		notifyPropertyChanged(BR.errorVisible);
	}

	private void handleSearchResult(Synset result) {
		searchResults = result.getTerms();

		notifyChange();
	}

	private void resetState() {
		searchResults = new ArrayList<>();
		errorMessage = null;
		notifyPropertyChanged(BR.errorVisible);
		notifyPropertyChanged(BR.searchResults);
	}

	@Bindable
	public String getSearchResults() {
		final List<String> terms = ListUtil.map(searchResults, t -> t.getTerm());
		return ListUtil.join(terms, '\n');
	}

	@Bindable
	public boolean isProgressBarVisible() {
		return isLoading;
	}

	@Bindable
	public boolean isErrorVisible() {
		return errorMessage != null;
	}

	@Bindable
	public String getErrorMessage() {
		return errorMessage;
	}

}

File: app/src/main/java/io/github/lmller/codetalks2017/SearchViewModel.kt
package io.github.lmller.codetalks2017

import android.databinding.BaseObservable
import android.databinding.Bindable
import io.reactivex.Observable
import io.reactivex.Scheduler
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import java.util.*
import java.util.concurrent.TimeUnit


class SearchViewModel(private val synonyms: Thesaurus) : BaseObservable() {

    private val logTag: String = synonyms.name()

    private var searchResultsTerms = emptyList<Term>()
        set(value) {
            field = value

            notifyChange()
        }

    private var isLoading = false
        set(value) {
            field = value
            notifyPropertyChanged(BR.progressBarVisible)
        }

    var errorMessage: String? = null
        @Bindable get
        private set(value) {
            field = value

            notifyPropertyChanged(BR.errorVisible)
        }

    val isErrorVisible: Boolean
        @Bindable
        get() = errorMessage != null

    val isProgressBarVisible: Boolean
        @Bindable get() = isLoading

    val searchResults: String
        @Bindable
        get() = searchResultsTerms.joinToString("\n") { term: Term -> term.term }

    fun observeTextChanges(queryTextChange: Observable<String>, scheduler: Scheduler = Schedulers.io()): Disposable {
        return queryTextChange
                .filter { text -> !text.isEmpty() }
                .debounce(250, TimeUnit.MILLISECONDS, scheduler)
                .flatMap { text ->
                    synonyms.findSynonym(text)
                            .doOnError { e ->
                                errorMessage = "$logTag responded with: ${e.message}"
                            }
                            .doOnSubscribe { _ -> resetState() }
                            .doOnTerminate { isLoading = false }
                            .onErrorReturnItem(Synset())
                }
                .subscribe { result ->
                    searchResultsTerms = result.terms
                }
    }

    private fun resetState() {
        searchResultsTerms = ArrayList()
        errorMessage = null
    }

}
